text/0000-crate-quarantine.md

```diff
@@ -0,0 +1,157 @@
+
+When a crate version is quarantined, a support ticket will also be opened within the crates.io help desk, thereby notifying the crates.io team of the newly quarantined version.
+
+### API
+
+A new field will be added to API endpoints that return crate versions called `published`. This will only be `false` for crate versions that are in quarantine.
+
+### Admin interface
+
+The crates.io admin console will be extended to provide the following new functionality:
+
+#### Rule definition
+
+A new CRUD area will be provided for rules to be defined and managed.
+
+#### Review
+
+A new review area will be added that lists crate versions currently in quarantine, along with options to download the `.crate` file and manage the version (ie by approving or rejecting the crate version).
+
+## `cargo publish`
+
+The wait loop in `cargo publish`[^wait-loop] currently refreshes the crate index and waits for the new crate version to appear.
+
+This will be extended to check the `quarantined` flag in the response; if present and `true`, `cargo publish` will not wait and will immediately report the warning(s) in the response to the user.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Project team member time
+
+The major drawback here is time: an actual human is going to need to be able to review any quarantined crate versions, which will require both expertise and availability. A crate being quarantined will almost certainly frustrate its owner(s), no matter what communication is provided, and the more time that is required to review the crate, the worse that experience will be.
+
+In the near term, it is unlikely that quarantine will be used in any day-to-day manner. The intention is to provide a break-glass-in-case-of-emergency option more than something that is in continuous use, but if this is used more frequently in normal operation, then workload will quickly become a concern.
+
+## Denial of service vectors
+
+This proposal addresses the needs of a Rust user — who doesn't want to inadvertently download and use a compromised or malicious crate — but does little to prevent denial of service on crates.io: both technically (since most publication steps will still take place for a quarantined crate version), and socially (since a human has to review the crate version).
+
+Overall, this is still an improvement in ecosystem security (since it can be used to prevent the index including malware), but it's not an all-encompassing magic solution to all our ecosystem security woes.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main alternative approach that would be available would be to, in essence, do nothing. Our current process is to handle malware and spam after crates are published, and if we needed to cut off a deluge of spam, we would be able to do so at the webserver level. It can be argued that this may be sufficient, given the relatively low level of incidents so far.
```

Comment:
```suggestion
The main alternative approach that would be available would be to, in essence, do nothing. Our current process is to handle malware and spam after crates are published, and if needed to cut off a deluge of spam, we have the ability to do so at the webserver level. It can be argued that this may be sufficient, given the relatively low level of incidents so far.
```